### ISSUE TITLE
System.BadImageFormatException: Bad IL Format when monitoring events on .NET Core 3.0-preview5

### ISSUE THREAD

**Issue by Wildenhaus** [fluentassertions/fluentassertions#1047]
### Description

FluentAssertions throws a `BadImageFormatException` when invoking a monitored event. This only occurs in the .NET Core 3.0 preview.

### Complete minimal example reproducing the issue

```csharp
[TestClass]
public class Repro
{
   public delegate void EventDelegate();

   public class Tester
   {
      public event EventDelegate TestEventDelegate;
      public event EventHandler TestEventHandler;

      public void RaiseEventDelegate()
      {
         TestEventDelegate?.Invoke();
      }

      public void RaiseEventHandler()
      {
         TestEventHandler?.Invoke( this, EventArgs.Empty );
      }
   }

   [TestMethod]
   public void TestRaiseEventDelegate()
   {
      var tester  = new Tester();
      var monitor = tester.Monitor();

      tester.RaiseEventDelegate();

      monitor.Should().Raise( nameof( Tester.TestEventDelegate ) );
   }

   [TestMethod]
   public void RaiseEventEventHandler()
   {
      var tester  = new Tester();
      var monitor = tester.Monitor();

      tester.RaiseEventHandler();

      monitor.Should().Raise( nameof( Tester.TestEventHandler ) );
   }
}
```

### Expected behavior:

`EventHandlerFactory` should emit a handler that does not cause a `BadImageFormatException` to be thrown.

### Actual behavior:

A `BadImageFormatException` is thrown when the event is invoked.

### Versions

* Which version of Fluent Assertions are you using?
FluentAssertions v5.6.0 via NuGet

* Which .NET runtime and version are you targeting?
.NET Core 3.0-preview5

### Additional Information

Have not looked into it. The exception is thrown immediately upon invoking the event. Bug does not occur in earlier versions of .NET Core (tested with 2.2).

**Comment by jnyrup** [comment]
I can confirm the crash.
* `netcoreapp2.2` passes the tests
* `netcoreapp3.0` fails the tests

```
Failed   TestRaiseEventDelegate
Error Message:
 Test method FA_1047.Repro.TestRaiseEventDelegate threw exception:
System.BadImageFormatException: Bad IL format.
Stack Trace:
    at EventDelegateDynamicHandler(IEventRecorder )
   at FA_1047.Repro.Tester.RaiseEventDelegate() in C:\Users\jonas\source\repos\FA-1047\FA-1047\Repro.cs:line 19
   at FA_1047.Repro.TestRaiseEventDelegate() in C:\Users\jonas\source\repos\FA-1047\FA-1047\Repro.cs:line 34

Failed   RaiseEventEventHandler
Error Message:
 Test method FA_1047.Repro.RaiseEventEventHandler threw exception:
System.BadImageFormatException: Bad IL format.
Stack Trace:
    at EventHandlerDynamicHandler(IEventRecorder , Object , EventArgs )
   at FA_1047.Repro.Tester.RaiseEventHandler() in C:\Users\jonas\source\repos\FA-1047\FA-1047\Repro.cs:line 24
   at FA_1047.Repro.RaiseEventEventHandler() in C:\Users\jonas\source\repos\FA-1047\FA-1047\Repro.cs:line 45
```

[FA-1047.zip](https://github.com/fluentassertions/fluentassertions/files/3210297/FA-1047.zip)

**Comment by jnyrup** [comment]
Got it working by altering `EventHandlerFactory.GenerateHandler` in one of two ways:

a) Change from `IEventRecorder` to `EventRecorder`
```c#
MethodInfo methodToCall = typeof(IEventRecorder).GetMethod("RecordEvent",
    BindingFlags.Instance | BindingFlags.Public);
```
into
```c#
MethodInfo methodToCall = typeof(EventRecorder).GetMethod("RecordEvent",
    BindingFlags.Instance | BindingFlags.Public);
```

or b) change from `OpCodes.Call` to `OpCodes.CallVirt`

```c#
ilGen.EmitCall(OpCodes.Call, methodToCall, null);
```
into
```c#
ilGen.EmitCall(OpCodes.CallVirt, methodToCall, null);
```

**Comment by dennisdoomen** [comment]
Did you find any documentation related to this? According to https://docs.microsoft.com/en-us/dotnet/api/system.reflection.emit.opcodes.call?view=netframework-4.8, it should all work. It could be a bug in the preview.

**Comment by jnyrup** [comment]
I didn't find any documentation, it was purely trial and error.

We should indeed cook up a minimal example and report it.
Either to know that:
* previously it worked by luck/coincidence,
* the is a change in .net core 3.0, or
* we've found a bug

**Comment by Wildenhaus** [comment]
Interface method calls will always be virtual, and the JIT can't inline those, so technically `Call` should have been `Callvirt` (or `IEventRecorder` should have been strongly typed) in the first place.

I'm not sure what has changed, but given the fix @jnyrup has found, I would assume that it has to do with more rigorous code checking on the part of RyuJIT.

**Comment by dennisdoomen** [comment]
Thanks @Wildenhaus. That's great information.

**Comment by dennisdoomen** [comment]
We probably need to add another test project that targets .NET Core 3.0

**Comment by lg2de** [comment]
> We probably need to add another test project that targets .NET Core 3.0

Is the build environment already supporting .NET Core 3?
We are facing this problem too.

**Comment by jnyrup** [comment]
According to https://github.com/appveyor/ci/issues/2907 VS2019 support is close to release.

Yet it doesn't show up at:
https://www.appveyor.com/docs/windows-images-software/#net-framework

The list above also `.NET Core 3.0.0-preview-27122-01 runtime` which corresponds to preview1 and to expose this bug, we need preview6.

@dennisdoomen I'm not familiar where to look for AppVeyor configuration or if I have access to that at all.

I tried adding a .NET Core 3.0 target locally and a few other tests breaks, mainly float point formatting as .NET Core 3.0 has more precise/correct floating point to string conversion.

I would be fine with:
* Fixing this issue in one PR, which can be included in a FA 5.7.0 release 
* Adding a .NET Core 3.0 target in a later PR

**Comment by dennisdoomen** [comment]
I've switched the AppVeyor build agent to Visual Studio 2019 Preview. Maybe this enough.

**Comment by jnyrup** [comment]
Fixed in #1057
